File: skills/bravado-data-api/scripts/sign.ts
/**
 * Bravado Data API request signing (HMAC-SHA256).
 * Node 18+, no dependencies. Copy into your project.
 *
 * const client = bravado({ publicKey, secret });
 * const top = await client.get("/trader-analytics/leaderboard", { window: "30d", limit: 25 });
 */
import { createHash, createHmac } from "node:crypto";

const EMPTY_BODY_SHA256 =
  "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";

/** RFC 3986 unreserved set: A-Z a-z 0-9 - _ . ~ ; everything else %XX uppercase. */
function encodeRfc3986(value: string): string {
  return encodeURIComponent(value).replace(
    /[!'()*]/g,
    (c) => "%" + c.charCodeAt(0).toString(16).toUpperCase(),
  );
}

/** Byte-order comparison. JS string compare is UTF-16 and diverges from UTF-8 above the BMP. */
function compareUtf8(a: string, b: string): number {
  return Buffer.compare(Buffer.from(a, "utf8"), Buffer.from(b, "utf8"));
}

export function canonicalPath(
  pathname: string,
  query: Record<string, unknown> = {},
): string {
  let path = decodeURIComponent(pathname);
  if (path.length > 1 && path.endsWith("/")) path = path.slice(0, -1);

  const entries = Object.entries(query)
    .filter(([, v]) => v !== undefined && v !== null)
    .map(([k, v]) => [k, String(v)] as [string, string]);

  if (entries.length === 0) return path;

  entries.sort((a, b) => compareUtf8(a[0], b[0]) || compareUtf8(a[1], b[1]));

  const qs = entries
    .map(([k, v]) => `${encodeRfc3986(k)}=${encodeRfc3986(v)}`)
    .join("&");
  return `${path}?${qs}`;
}

export function signHeaders(opts: {
  publicKey: string;
  secret: string;
  method: string;
  pathname: string;
  query?: Record<string, unknown>;
  rawBody?: Buffer | string;
  timestampMs?: number;
}): Record<string, string> {
  const timestamp = String(opts.timestampMs ?? Date.now());
  const method = opts.method.toUpperCase();
  const path = canonicalPath(opts.pathname, opts.query ?? {});

  const bodyHash =
    opts.rawBody === undefined || opts.rawBody === ""
      ? EMPTY_BODY_SHA256
      : createHash("sha256").update(opts.rawBody).digest("hex");

  const payload = [timestamp, method, path, bodyHash].join("\n");
  const signature = createHmac("sha256", opts.secret)
    .update(payload)
    .digest("hex");

  return {
    "X-BRAVADO-API-KEY": opts.publicKey,
    "X-BRAVADO-TIMESTAMP": timestamp,
    "X-BRAVADO-SIGNATURE": signature,
  };
}

export function bravado(config: {
  publicKey: string;
  secret: string;
  host?: string;
}) {
  const host = config.host ?? "https://api.bravado.io";

  return {
    async get<T = unknown>(
      pathname: string,
      query: Record<string, unknown> = {},
    ): Promise<T> {
      const headers = signHeaders({
        publicKey: config.publicKey,
        secret: config.secret,
        method: "GET",
        pathname,
        query,
      });

      // The canonical form is for SIGNING. The wire URL keeps the original
      // pathname (the server decodes it once) plus the canonical querystring,
      // so what the server re-derives matches what we signed.
      const canonical = canonicalPath(pathname, query);
      const qs = canonical.includes("?") ? canonical.slice(canonical.indexOf("?")) : "";
      const res = await fetch(host + pathname + qs, { headers });
      if (!res.ok) {
        const body = await res.text();
        throw new Error(`bravado ${res.status}: ${body}`);
      }
      return (await res.json()) as T;
    },
  };
}

File: skills/bravado-data-api/scripts/sign.py
"""
Bravado Data API request signing (HMAC-SHA256).
Python 3.9+, stdlib only. Copy into your project.

    client = Bravado(public_key, secret)
    top = client.get("/trader-analytics/leaderboard", {"window": "30d", "limit": 25})
"""
from __future__ import annotations

import hashlib
import hmac
import json
import time
import urllib.parse
import urllib.request

EMPTY_BODY_SHA256 = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
UNRESERVED = "-_.~"


def _encode(value: str) -> str:
    """RFC 3986 unreserved set; everything else %XX uppercase (quote() already uppercases)."""
    return urllib.parse.quote(value, safe=UNRESERVED)


def canonical_path(pathname: str, query: dict | None = None) -> str:
    path = urllib.parse.unquote(pathname)
    if len(path) > 1 and path.endswith("/"):
        path = path[:-1]

    items = [
        (str(k), str(v).lower() if isinstance(v, bool) else str(v))
        for k, v in (query or {}).items()
        if v is not None
    ]
    if not items:
        return path

    # Byte order, not codepoint order.
    items.sort(key=lambda kv: (kv[0].encode("utf-8"), kv[1].encode("utf-8")))
    qs = "&".join(f"{_encode(k)}={_encode(v)}" for k, v in items)
    return f"{path}?{qs}"


def sign_headers(
    public_key: str,
    secret: str,
    method: str,
    pathname: str,
    query: dict | None = None,
    raw_body: bytes | None = None,
    timestamp_ms: int | None = None,
) -> dict:
    timestamp = str(timestamp_ms if timestamp_ms is not None else int(time.time() * 1000))
    path = canonical_path(pathname, query)
    body_hash = (
        EMPTY_BODY_SHA256 if not raw_body else hashlib.sha256(raw_body).hexdigest()
    )

    payload = "\n".join([timestamp, method.upper(), path, body_hash])
    signature = hmac.new(
        secret.encode("utf-8"), payload.encode("utf-8"), hashlib.sha256
    ).hexdigest()

    return {
        "X-BRAVADO-API-KEY": public_key,
        "X-BRAVADO-TIMESTAMP": timestamp,
        "X-BRAVADO-SIGNATURE": signature,
    }


class Bravado:
    def __init__(self, public_key: str, secret: str, host: str = "https://api.bravado.io"):
        self.public_key = public_key
        self.secret = secret
        self.host = host

    def get(self, pathname: str, query: dict | None = None):
        headers = sign_headers(self.public_key, self.secret, "GET", pathname, query)
        # Canonical form is for SIGNING only. The wire URL keeps the original
        # pathname (the server decodes it once) plus the canonical querystring.
        canonical = canonical_path(pathname, query)
        qs = canonical[canonical.index("?"):] if "?" in canonical else ""
        req = urllib.request.Request(self.host + pathname + qs, headers=headers)
        with urllib.request.urlopen(req) as res:
            return json.loads(res.read())
